File: python/pov/pov.py
from json import dumps
from copy import deepcopy

class Tree:
    def __init__(self, label, children=None):
        self.label = label
        # Ensure children is always a list, even if None is passed
        self.children = children if children is not None else []

    def __dict__(self):
        # Sort children by label for consistent dictionary representation and comparison
        # Use a lambda function for sorting based on the 'label' attribute
        return {self.label: [c.__dict__() for c in sorted(self.children, key=lambda node: node.label)]}

    def __str__(self, indent=None):
        # Use the custom __dict__ method for JSON serialization
        return dumps(self.__dict__(), indent=indent)

    def __lt__(self, other):
        # Comparison based on label
        return self.label < other.label

    def __eq__(self, other):
        # Check if the other object is an instance of Tree before comparing dictionaries
        if not isinstance(other, Tree):
            return NotImplemented # Indicate comparison is not supported
        # Use the custom __dict__ method for equality comparison
        return self.__dict__() == other.__dict__()

    def __getstate__(self):
        # Return the state needed to reconstruct the object
        # Ensure children are part of the state
        return {'label': self.label, 'children': self.children}

    def __setstate__(self, state):
        # Restore the object's state from the dictionary
        self.label = state['label']
        self.children = state['children']

    def _find_node_with_path(self, target_label, current_path_nodes=None):
        """
        Helper method to find a node by its label using DFS.
        Returns the node object and the path (list of node objects) from the root to that node.
        Returns (None, []) if the node is not found.
        """
        if current_path_nodes is None:
            current_path_nodes = []

        # Create a new list for the path at this level to avoid modifying parent's path
        path_here = current_path_nodes + [self]

        if self.label == target_label:
            return self, path_here # Node found

        # Recursively search in children
        for child in self.children:
            # Pass a copy of path_here to avoid modification by sibling branches
            found_node, found_path = child._find_node_with_path(target_label, path_here)
            if found_node:
                return found_node, found_path # Propagate the found result upwards

        # Node not found in this subtree
        return None, []

    def from_pov(self, from_node_label):
        """
        Re-roots the tree from the perspective of the node with the given label.
        Returns a new Tree instance representing the re-rooted tree.
        Raises ValueError if the specified node is not found.
        """
        original_root = self
        node_to_reroot, path_nodes = original_root._find_node_with_path(from_node_label)

        if not node_to_reroot:
            raise ValueError("Tree could not be reoriented")

        # If the target node is the original root, return a deep copy of the original tree
        if node_to_reroot == original_root:
            return deepcopy(original_root)

        # --- Rerooting Logic (Path Modification Approach) ---
        # Create a deep copy of the entire tree first to work on.
        # This ensures all nodes and their relationships are copied initially.
        copied_root = deepcopy(original_root)

        # Find the path again, but this time within the *copied* tree.
        # We need the actual node objects from the copy.
        node_to_reroot_copy, path_nodes_copy = copied_root._find_node_with_path(from_node_label)

        if not node_to_reroot_copy:
             # This should not happen if the original find worked, but safety check.
             raise ValueError("Error finding node in copied tree.")

        # Iterate up the path in the *copied* tree, modifying relationships
        # Start from the node *above* the target node (its parent) up to the root.
        for i in range(len(path_nodes_copy) - 1, 0, -1):
            parent_copy = path_nodes_copy[i-1]
            child_copy = path_nodes_copy[i] # This is the node that needs to become the parent

            # Remove child_copy from parent_copy's children list
            # We need to compare objects, not just labels, in case of duplicate labels.
            parent_copy.children = [c for c in parent_copy.children if c is not child_copy]

            # Add parent_copy as a child to child_copy
            child_copy.children.append(parent_copy)

        # The copy of the original target node is the new root of the reoriented tree.
        return node_to_reroot_copy


    def path_to(self, from_node_label, to_node_label):
        """
        Finds the shortest path between two nodes identified by their labels.
        Returns a list of labels representing the path.
        Raises ValueError if either node is not found.
        """
        original_root = self
        _, path1_nodes = original_root._find_node_with_path(from_node_label)
        # Check if the source node was found
        if not path1_nodes:
            # Test expects "Tree could not be reoriented" for missing source node
            raise ValueError("Tree could not be reoriented")

        _, path2_nodes = original_root._find_node_with_path(to_node_label)
        # Check if the destination node was found
        if not path2_nodes:
            # Test expects "No path found" for missing destination node
            raise ValueError("No path found")

        # Find the index of the Lowest Common Ancestor (LCA) node in the paths
        lca_index = 0
        min_len = min(len(path1_nodes), len(path2_nodes))
        # Increment lca_index as long as nodes at that index match in both paths
        while lca_index < min_len and path1_nodes[lca_index].label == path2_nodes[lca_index].label:
            lca_index += 1
        # Decrement to get the index of the last common node (the LCA)
        lca_index -= 1

        # Construct the path: from_node -> ... -> LCA -> ... -> to_node

        # Part 1: Path from from_node up to LCA
        # Slice path1 from the LCA index onwards
        path_up_nodes = path1_nodes[lca_index:]
        # Get labels and reverse to get the order from_node -> ... -> LCA
        path_up_labels = [node.label for node in path_up_nodes]
        path_up_labels.reverse()

        # Part 2: Path from LCA down to to_node
        # Slice path2 from the node *after* the LCA down to the target node
        path_down_nodes = path2_nodes[lca_index + 1:]
        path_down_labels = [node.label for node in path_down_nodes]

        # Combine the paths: (from_node -> LCA) + (LCA -> to_node, excluding LCA)
        final_path = path_up_labels + path_down_labels
        return final_path

File: go/pov/pov.go
package pov

import "sort" // Import sort for String() method consistency

// Tree represents a node in a tree graph.
type Tree struct {
	value    string
	children []*Tree
	parent   *Tree // Adding parent pointer to facilitate reparenting
}

// New creates and returns a new Tree with the given root value and children.
func New(value string, children ...*Tree) *Tree {
	tr := &Tree{
		value:    value,
		children: children,
	}
	// Set the parent pointer for each child
	for _, child := range children {
		if child != nil { // Ensure child is not nil before setting parent
			child.parent = tr
		}
	}
	return tr
}

// Value returns the value at the root of a tree.
func (tr *Tree) Value() string {
	if tr == nil {
		return "" // Consistent with potential test expectations for nil tree
	}
	return tr.value
}

// Children returns a slice containing the children of a tree.
func (tr *Tree) Children() []*Tree {
	if tr == nil {
		return nil
	}
	// Return a copy to prevent external modification? Let's stick to direct slice for now.
	// If tests fail due to modification of returned slice, create a copy here.
	childrenCopy := make([]*Tree, len(tr.children))
	copy(childrenCopy, tr.children)
	return childrenCopy // Returning a copy is safer
}

// String describes a tree in a compact S-expression format.
// Added sorting for deterministic output.
func (tr *Tree) String() string {
	if tr == nil {
		return "nil"
	}
	result := tr.Value()
	children := tr.Children() // Use the method to get children
	if len(children) == 0 {
		return result
	}
	// Sort children by value for consistent string representation
	sort.Slice(children, func(i, j int) bool {
		// Handle potential nil children if they can occur
		if children[i] == nil { return true }
		if children[j] == nil { return false }
		return children[i].Value() < children[j].Value()
	})
	result = "(" + result
	for _, ch := range children {
		result += " " + ch.String()
	}
	result += ")"
	return result
}


// --- Helper Functions ---

// findNodeAndPath uses BFS to find a node by value and returns the node
// and the path (list of nodes) from the given 'tr' (assumed root for this search) to that node.
func findNodeAndPath(tr *Tree, value string) (*Tree, []*Tree) {
	if tr == nil {
		return nil, nil
	}

	queue := []*Tree{tr}
	// parentMap stores the parent of each node *in the context of this BFS traversal*
	parentMap := make(map[*Tree]*Tree)
	parentMap[tr] = nil // Mark root's parent as nil for path reconstruction termination

	visited := make(map[*Tree]bool)
	visited[tr] = true

	var targetNode *Tree = nil

	// BFS traversal
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		if current.value == value {
			targetNode = current
			break // Found the target node
		}

		// Explore children (use the Children() method which returns a safe copy)
		for _, child := range current.Children() { // Use Children() method
			if child != nil && !visited[child] {
				visited[child] = true
				parentMap[child] = current // Record parent for path reconstruction
				queue = append(queue, child)
			}
		}
	}

	if targetNode == nil {
		return nil, nil // Node not found
	}

	// Reconstruct the path from targetNode back to the root (tr)
	path := []*Tree{}
	curr := targetNode
	for curr != nil {
		path = append(path, curr)
		// Move up using the parentMap created during BFS
		curr = parentMap[curr]
	}

	// Reverse the path to be root -> targetNode
	for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
		path[i], path[j] = path[j], path[i]
	}

	return targetNode, path
}

// findNodeAnywhere uses BFS starting from startNode and explores both children and parent pointers
// to find a node anywhere in the connected component.
func findNodeAnywhere(startNode *Tree, value string) *Tree {
    if startNode == nil {
        return nil
    }
    queue := []*Tree{startNode}
    visited := make(map[*Tree]bool)
    visited[startNode] = true

    for len(queue) > 0 {
        current := queue[0]
        queue = queue[1:]

        if current.value == value {
            return current
        }

        // Explore children (using direct access for internal traversal)
        for _, child := range current.children {
            if child != nil && !visited[child] {
                visited[child] = true
                queue = append(queue, child)
            }
        }
        // Explore parent
        if current.parent != nil && !visited[current.parent] {
            visited[current.parent] = true
            queue = append(queue, current.parent)
        }
    }
    return nil // Node not found in the connected component
}


// --- POV problem-specific functions ---

// FromPov returns the pov from the node specified in the argument by modifying the tree structure.
func (tr *Tree) FromPov(from string) *Tree {
	// Find the target node anywhere in the tree structure connected to tr
	targetNode := findNodeAnywhere(tr, from)
	if targetNode == nil {
		return nil // Node not found
	}

	// Reparenting logic: reverse path from targetNode up towards the original root/boundaries.
	var prev *Tree = nil // Represents the new child (previously the parent)
	curr := targetNode

	for curr != nil {
		originalParent := curr.parent // Store original parent before modification

		// 1. Set current node's parent pointer to the node processed in the previous iteration (its new parent)
		curr.parent = prev

		// 2. Modify current node's children list:
		newChildren := []*Tree{}
		// Keep all original children *except* the one that is becoming the new parent (`prev`)
		for _, child := range curr.children {
			if child != prev {
				newChildren = append(newChildren, child)
			}
		}
		// Add the original parent as a child, *if* it exists
		if originalParent != nil {
			// Check if originalParent is already in newChildren (shouldn't be if prev logic is correct)
			isAlreadyChild := false
			for _, c := range newChildren {
				if c == originalParent {
					isAlreadyChild = true
					break
				}
			}
			if !isAlreadyChild {
				newChildren = append(newChildren, originalParent)
			}
		}
		curr.children = newChildren

		// Move up the original path: current becomes previous, original parent becomes current
		prev = curr
		curr = originalParent
	}

	// targetNode is now the root of the reparented tree
	if targetNode != nil {
		targetNode.parent = nil // Ensure the new root's parent is explicitly nil
	}
	return targetNode
}


// PathTo returns the shortest path between two nodes in the tree using LCA. Does not modify the tree.
func (tr *Tree) PathTo(from, to string) []string {
	if from == to {
		// Check if the node exists before returning path of length 1
		node, _ := findNodeAndPath(tr, from) // Use findNodeAndPath which searches downwards from tr
		if node != nil {
			return []string{from}
		} else {
			// If not found downwards, it might still exist elsewhere if tr wasn't the absolute root.
			// However, PathTo is usually called on the root. Let's assume findNodeAndPath is sufficient.
			return nil // Node doesn't exist or isn't reachable downwards from tr
		}
	}

	// Find nodes and paths from the root (tr)
	fromNode, pathFromRootToFrom := findNodeAndPath(tr, from)
	if fromNode == nil {
		return nil // 'from' node not found downwards from tr
	}

	toNode, pathFromRootToTo := findNodeAndPath(tr, to)
	if toNode == nil {
		return nil // 'to' node not found downwards from tr
	}

	// Find the Lowest Common Ancestor (LCA)
	lcaIndex := -1
	maxLength := len(pathFromRootToFrom)
	if len(pathFromRootToTo) < maxLength {
		maxLength = len(pathFromRootToTo)
	}
	for i := 0; i < maxLength; i++ {
		if pathFromRootToFrom[i] == pathFromRootToTo[i] {
			lcaIndex = i
		} else {
			break // Divergence point found
		}
	}

	if lcaIndex == -1 {
		// This implies tr itself wasn't a common ancestor, which is unexpected.
		return nil // Error or unexpected tree structure
	}

	// Construct the path: from -> ... -> LCA -> ... -> to
	pathResult := []string{}

	// Part 1: Path from 'from' up to LCA
	// Iterate backwards on pathFromRootToFrom from 'fromNode' down to the node *at* LCA
	for i := len(pathFromRootToFrom) - 1; i >= lcaIndex; i-- {
		pathResult = append(pathResult, pathFromRootToFrom[i].value)
	}

	// Part 2: Path from LCA (exclusive) down to 'to'
	// Iterate forwards on pathFromRootToTo from the node *after* LCA to 'toNode'
	for i := lcaIndex + 1; i < len(pathFromRootToTo); i++ {
		pathResult = append(pathResult, pathFromRootToTo[i].value)
	}

	return pathResult
}
